fabric-data-generation-api-v1/src/main/java/net/fabricmc/fabric/api/datagen/v1/provider/FabricDynamicRegistryProvider.java

```diff
@@ -227,21 +274,31 @@ private <T> CompletableFuture<?> writeRegistryEntries(DataWriter writer, Registr
 		final DataOutput.PathResolver pathResolver = output.getResolver(DataOutput.OutputType.DATA_PACK, directoryName);
 		final List<CompletableFuture<?>> futures = new ArrayList<>();
 
-		for (Map.Entry<RegistryKey<T>, T> entry : entries.entries.entrySet()) {
+		for (Map.Entry<RegistryKey<T>, EntryWithConditions<T>> entry : entries.entries.entrySet()) {
 			Path path = pathResolver.resolveJson(entry.getKey().getValue());
-			futures.add(writeToPath(path, writer, ops, entries.elementCodec, entry.getValue()));
+			futures.add(writeToPath(path, writer, ops, entries.elementCodec, entry.getValue().value(), entry.getValue().conditions()));
 		}
 
 		return CompletableFuture.allOf(futures.toArray(CompletableFuture[]::new));
 	}
 
-	private static <E> CompletableFuture<?> writeToPath(Path path, DataWriter cache, DynamicOps<JsonElement> json, Encoder<E> encoder, E value) {
+	private static <E> CompletableFuture<?> writeToPath(Path path, DataWriter cache, DynamicOps<JsonElement> json, Encoder<E> encoder, E value, @Nullable ResourceCondition[] conditions) {
 		Optional<JsonElement> optional = encoder.encodeStart(json, value).resultOrPartial((error) -> {
 			LOGGER.error("Couldn't serialize element {}: {}", path, error);
 		});
 
 		if (optional.isPresent()) {
-			return DataProvider.writeToPath(cache, optional.get(), path);
+			JsonElement jsonElement = optional.get();
+
+			if (conditions != null && conditions.length > 0) {
+				if (!jsonElement.isJsonObject()) {
+					LOGGER.error("Cannot add conditions to {}: JSOn is a non-object value", path);
```

Comment:
```suggestion
					LOGGER.error("Cannot add conditions to {}: JSON is a non-object value", path);
```

Reply:
I wonder if this would be better off to throw? I dont see a reason not to.

Reply:
Yeah, I based this off the fact that `writeToPath` logs an error if it cannot serialise, but maybe that should throw too?

Reply:
Humm, I see. I dont think we should change DataProvider to throw (even though I think it should). But we should throw here, its much better to force the mod dev to fix their issue than lead to possible issues at runtime.

Reply:
Oh, I meant this one within `FabricDynamicRegistryProvider.writeToPath`:

https://github.com/FabricMC/fabric/blob/1997dc55222c6bc9183dbf592672a56189a24a14/fabric-data-generation-api-v1/src/main/java/net/fabricmc/fabric/api/datagen/v1/provider/FabricDynamicRegistryProvider.java#L239-L241

Reply:
Oh right, this does match vanilla, I dont think there is a strong need to change existing stuff as part of this PR.

Im happy either way with the PR, ill let you choose what you think is best.